src/host.rs

```diff
@@ -97,29 +255,81 @@ impl Host {
 
     pub(crate) fn recv_from(&mut self, peer: SocketAddr) -> (Option<Envelope>, Rc<Notify>) {
         let now = Instant::now();
-        let notify = self.notify.clone();
 
         let deque = self.inbox.entry(peer).or_default();
+        let notify = self.notify.clone();
 
         match deque.front() {
             Some(Envelope {
                 instructions: DeliveryInstructions::DeliverAt(time),
                 ..
             }) if *time <= now => {
-                self.version += 1;
-                (deque.pop_front(), notify)
+                let ret = (deque.pop_front(), notify);
+                self.bump_version();
+                ret
             }
             _ => (None, notify),
         }
     }
 
+    pub(crate) fn recv_on(&mut self, pair: SocketPair) -> Option<Segment> {
+        let now = Instant::now();
+
+        let deque = self
+            .connections
+            .get_mut(&pair)
+            .map(|i| &mut i.deque)
+            .expect("no connection");
+
+        match deque.front() {
+            Some(StreamEnvelope {
+                instructions: DeliveryInstructions::DeliverAt(time),
+                ..
+            }) if *time <= now => {
+                let ret = deque.pop_front().map(|e| e.segment);
+                self.bump_version();
+                ret
+            }
+            _ => None,
+        }
+    }
+
     /// Releases all messages previously received from [`peer`]. These messages
     /// may be received immediately (on the next call to `[Host::recv]`).
     pub(crate) fn release(&mut self, peer: SocketAddr) {
         let now = Instant::now();
-        let deque = self.inbox.entry(peer).or_default();
 
-        for envelope in deque {
+        if let Some(listener) = self.listener.as_mut() {
```

Comment:
I don't love that we have to release from two places. I'm not worried about the small amount of code duplication, but I do think it's a smell that the abstractions aren't quite right. All internal though..

Reply:
Agree.